kclvm/query/src/tests.rs

```diff
@@ -216,3 +216,220 @@ fn test_parse_property_path() {
     assert!(parse_attribute_path(r#"a.[b.c-d.e"#).is_err(),);
     assert!(parse_attribute_path(r#"a.[b.c]-d.e"#).is_err(),);
 }
+
+#[test]
+fn test_list_variables() {
+    let file = PathBuf::from("./src/test_data/test_list_variables/supported.k")
+        .canonicalize()
+        .unwrap()
+        .display()
+        .to_string();
+    let test_cases = vec![
+        ("a", "1"),
+        ("a1", "2"),
+        ("a3", "3m"),
+        ("b1", "True"),
+        ("b2", "False"),
+        ("s1", "\"Hello\""),
+        ("array1", "[1, 2, 3]"),
+        ("dict1", "{\"a\": 1, \"b\": 2}"),
+        ("dict1.a", "1"),
+        ("dict1.b", "2"),
+        (
+            "dict2",
+            r#"{
+    "a": 1
+    "b": {
+        "c": 2
+        "d": 3
+    }
+}"#,
+        ),
+        ("dict2.a", "1"),
+        (
+            "dict2.b",
+            r#"{
+    "c": 2
+    "d": 3
+}"#,
+        ),
+        ("dict2.b.c", "2"),
+        ("dict2.b.d", "3"),
+        (
+            "sha",
+            r#"A {
+    name: "Hello"
+    ids: [1, 2, 3]
+    data: {
+        "a": {
+            "b": {"c": 2}
+        }
+    }
+}"#,
+        ),
+        ("sha.name", "\"Hello\""),
+        ("sha.ids", "[1, 2, 3]"),
+        (
+            "sha.data",
+            r#"{
+    "a": {
+        "b": {"c": 2}
+    }
+}"#,
+        ),
+        (
+            "sha.data.a",
+            r#"{
+    "b": {"c": 2}
+}"#,
+        ),
+        ("sha.data.a.b", r#"{"c": 2}"#),
+        ("sha.data.a.b.c", "2"),
+        (
+            "shb",
+            r#"B {
+    a: {
+        name: "HelloB"
+        ids: [4, 5, 6]
+        data: {
+            "d": {
+                "e": {"f": 3}
+            }
+        }
+    }
+}"#,
+        ),
+        (
+            "shb.a",
+            r#"{
+    name: "HelloB"
+    ids: [4, 5, 6]
+    data: {
+        "d": {
+            "e": {"f": 3}
+        }
+    }
+}"#,
+        ),
+        ("shb.a.name", "\"HelloB\""),
+        ("shb.a.ids", "[4, 5, 6]"),
+        (
+            "shb.a.data",
+            r#"{
+    "d": {
+        "e": {"f": 3}
+    }
+}"#,
+        ),
+        (
+            "shb.a.data.d",
+            r#"{
+    "e": {"f": 3}
+}"#,
+        ),
+        ("shb.a.data.d.e", "{\"f\": 3}"),
+        ("uconfa.name", "\"b\""),
+        ("c.a", "{ids: [7, 8, 9]}"),
+    ];
+
+    for (spec, expected) in test_cases {
+        let specs = vec![spec.to_string()];
+        let result = list_variables(file.clone(), specs).unwrap();
+        assert_eq!(result.select_result.get(spec).unwrap(), expected);
+    }
+}
+
+#[test]
+fn test_list_all_variables() {
+    let file = PathBuf::from("./src/test_data/test_list_variables/supported.k")
+        .canonicalize()
+        .unwrap()
+        .display()
+        .to_string();
+    let test_cases = vec![
```

Comment:
Add more test cases for the complex expression. e.g., `"{}/{}".format(registry, image)`